File: src/2023/day1.zig
const std = @import("std");
const utils = @import("utils");

pub fn main() !void {
    // https://adventofcode.com/2023/day/1
    utils.printTitle(2023, 1, "Trebuchet?!");

    const m = utils.Monitor.init();
    defer m.deinit();

    @setEvalBranchQuota(1500000);

    const input = comptime utils.trimInput(@embedFile("input"));
    utils.printAnswer(1, comptime part1(input));
    utils.printAnswer(2, comptime part2(input));
}

fn getDigit(text: []const u8) ?i32 {
    if (text[0] >= '0' and text[0] <= '9') {
        return text[0] - '0';
    } else if (std.mem.startsWith(u8, text, "one")) {
        return 1;
    } else if (std.mem.startsWith(u8, text, "two")) {
        return 2;
    } else if (std.mem.startsWith(u8, text, "three")) {
        return 3;
    } else if (std.mem.startsWith(u8, text, "four")) {
        return 4;
    } else if (std.mem.startsWith(u8, text, "five")) {
        return 5;
    } else if (std.mem.startsWith(u8, text, "six")) {
        return 6;
    } else if (std.mem.startsWith(u8, text, "seven")) {
        return 7;
    } else if (std.mem.startsWith(u8, text, "eight")) {
        return 8;
    } else if (std.mem.startsWith(u8, text, "nine")) {
        return 9;
    }

    return null;
}

fn findFirstDigit(text: []const u8) i32 {
    for (text, 0..) |_, i| {
        if (getDigit(text[i..])) |digit| {
            return digit;
        }
    }

    return 0;
}

fn findLastDigit(text: []const u8) i32 {
    var i: usize = text.len;
    while (i > 0) {
        i -= 1;
        if (getDigit(text[i..])) |digit| {
            return digit;
        }
    }

    return 0;
}

fn part1(text: []const u8) i32 {
    const textTrimmed = std.mem.trim(u8, text, "\r\n");
    var it = std.mem.tokenizeAny(u8, textTrimmed, "\r\n");

    var sum: i32 = 0;
    while (it.next()) |line| {
        const first = std.mem.indexOfAny(u8, line, "123456789").?;
        const last = std.mem.lastIndexOfAny(u8, line, "123456789").?;
        const num = (line[first] - '0') * 10 + (line[last] - '0');
        sum += num;
    }

    return sum;
}

fn part2(text: []const u8) i32 {
    const textTrimmed = std.mem.trim(u8, text, "\r\n");
    var it = std.mem.tokenizeAny(u8, textTrimmed, "\r\n");
    var sum: i32 = 0;
    while (it.next()) |line| {
        const first = findFirstDigit(line);
        const last = findLastDigit(line);
        sum += first * 10 + last;
    }

    return sum;
}

test "part 1" {
    const example =
        \\1abc2
        \\pqr3stu8vwx
        \\a1b2c3d4e5f
        \\treb7uchet
    ;
    try std.testing.expectEqual(142, comptime part1(example));
}

test "part 2" {
    const example =
        \\two1nine
        \\eightwothree
        \\abcone2threexyz
        \\xtwone3four
        \\4nineeightseven2
        \\zoneight234
        \\7pqrstsixteen
    ;
    @setEvalBranchQuota(2000);
    try std.testing.expectEqual(281, comptime part2(example));
}

File: src/2023/day1.cpp
#include "day1.hpp"

#include "../common.hpp"

using namespace std::literals;

static auto Part1(std::string_view lines)
{
    int total = 0;

    for (auto line : Split(lines, '\n'))
    {
        char first = '\0';
        char last = '\0';

        for (char c : line)
        {
            if (isdigit(c))
            {
                if (first == '\0')
                {
                    first = c;
                }

                last = c;
            }
        }

        total += svtoi<int>(std::format("{}{}", first, last));
    }

    return total;
}

static int GetDigit(std::string_view line)
{
    switch (char const c = line.front(); c)
    {
        case '0':
            throw std::runtime_error{"invalid digit"};
        case '1':
        case '2':
        case '3':
        case '4':
        case '5':
        case '6':
        case '7':
        case '8':
        case '9':
            return c - '0';
        case 'o':
            if (line.starts_with("one"sv))
            {
                return 1;
            }
            break;
        case 't':
            if (line.starts_with("two"sv))
            {
                return 2;
            }
            else if (line.starts_with("three"sv))
            {
                return 3;
            }
            break;
        case 'f':
            if (line.starts_with("four"sv))
            {
                return 4;
            }
            else if (line.starts_with("five"sv))
            {
                return 5;
            }
            break;
        case 's':
            if (line.starts_with("six"sv))
            {
                return 6;
            }
            else if (line.starts_with("seven"sv))
            {
                return 7;
            }
            break;
        case 'e':
            if (line.starts_with("eight"sv))
            {
                return 8;
            }
            break;
        case 'n':
            if (line.starts_with("nine"sv))
            {
                return 9;
            }
            break;
    }

    return 0;
}

static auto GetCalibration2(std::string_view line)
{
    trim(line);
    std::vector<int> digits;

    while (true)
    {
        auto const digit = GetDigit(line);
        if (digit != 0)
        {
            digits.push_back(digit);
        }

        line = line.substr(1);

        if (line.empty())
        {
            break;
        }
    }

    return svtoi<int64_t>(std::format("{}{}", digits.front(), digits.back()));
}

static auto Part2(std::string_view lines)
{
    int64_t total = 0;

    for (auto line : Split(lines, '\n'))
    {
        total += GetCalibration2(line);
    }

    return total;
}

int main()
{
    // https://adventofcode.com/2023/day/1
    std::println("Day 1, 2023: Trebuchet?!");

    Assert(142 == Part1(example::part1));
    Assert(281 == Part2(example::part2));
    Assert(29 == GetCalibration2(Split(example::part2, '\n')[0]));
    Assert(83 == GetCalibration2(Split(example::part2, '\n')[1]));
    Assert(13 == GetCalibration2(Split(example::part2, '\n')[2]));
    Assert(24 == GetCalibration2(Split(example::part2, '\n')[3]));
    Assert(42 == GetCalibration2(Split(example::part2, '\n')[4]));
    Assert(14 == GetCalibration2(Split(example::part2, '\n')[5]));
    Assert(76 == GetCalibration2(Split(example::part2, '\n')[6]));

    auto const part1 = Part1(GetInput());
    std::println("  Part 1: {}", part1);
    Assert(54'632 == part1);

    auto const part2 = Part2(GetInput());
    std::println("  Part 2: {}", part2);
    Assert(54'019 == part2);
}
